app/upgrades.go

```diff
@@ -230,6 +232,41 @@ func removeP8eMemorializeContractFee(ctx sdk.Context, app *App) {
 	}
 }
 
+// removeInactiveValidators unbonds all delegations from inactive validators, triggering their removal from the validator set.
+func removeInactiveValidators(ctx sdk.Context, app *App) {
+	unbondingTimeParam := app.StakingKeeper.GetParams(ctx).UnbondingTime
+	ctx.Logger().Info(fmt.Sprintf("removing any validator that has been inactive (unbonded) for %d hours", int64(unbondingTimeParam.Hours()/24)))
+	removalCount := 0
+	validators := app.StakingKeeper.GetAllValidators(ctx)
+	for _, validator := range validators {
+		if validator.IsUnbonded() {
+			inActiveDuration := ctx.BlockTime().Sub(validator.UnbondingTime)
+			if inActiveDuration >= unbondingTimeParam {
+				ctx.Logger().Info(fmt.Sprintf("validator %v has been inactive (unbonded) for %d days and will be removed", validator.OperatorAddress, int64(inActiveDuration.Hours()/24)))
+				valAddress, err := sdk.ValAddressFromBech32(validator.OperatorAddress)
+				if err != nil {
+					panic(err)
+				}
+				delegations := app.StakingKeeper.GetValidatorDelegations(ctx, valAddress)
+				for _, delegation := range delegations {
+					ctx.Logger().Info(fmt.Sprintf("undelegate delegator %v from validator %v of all shares (%v)", delegation.DelegatorAddress, validator.OperatorAddress, delegation.GetShares()))
+					// unbonding all of a delegators shares, method will remove validator from set once all have been removed.
+					_, err := app.StakingKeeper.Undelegate(ctx, delegation.GetDelegatorAddr(), valAddress, delegation.GetShares())
+					if err != nil {
+						panic(err)
+					}
+				}
+				// this should probably never be true, an unbonded delegator will be removed when the last delegator has called Unbond
+				if len(delegations) == 0 && validator.DelegatorShares.IsZero() {
+					app.StakingKeeper.RemoveValidator(ctx, valAddress)
+				}
+				removalCount++
+			}
+		}
+	}
+	ctx.Logger().Info(fmt.Sprintf("a total of %d inactive (unbonded) validators have been removed", removalCount))
```

Comment:
more correctly this code isn't intended to "remove unbonded validators" ... but to unbond delegations from stale/unbonded validators.

Reply:
Ok, I changed the text on that along with a few other refactors to keep things consistent.